add langgraph oneagent demo with openpipeline secret redaction (#149)
dynatrace-oss/dynatrace-ai-agent-instrumentation-examples: langgraph/oneagent/README.md

# LangGraph + OneAgent

> This sample traces a [LangGraph](https://langchain-ai.github.io/langgraph/) agent with Dynatrace using **OneAgent auto-instrumentation** — no manual OpenTelemetry export or collector. OneAgent instruments the underlying Azure OpenAI SDK calls the graph makes and ships `gen_ai.*` spans straight to Dynatrace.

Reviewer: I would omit some low-level details from the intro
```suggestion
This sample traces a [LangGraph](https://langchain-ai.github.io/langgraph/) agent with Dynatrace using **OneAgent auto-instrumentation** — no manual OpenTelemetry export or collector. OneAgent instruments the underlying Azure OpenAI SDK calls and ships `gen_ai.*` spans straight to Dynatrace.
```

For a collector-based variant (which exports via OTLP and can redact secrets before they leave the host), see [`langgraph/opentelemetry`](../opentelemetry).

## What this sample does

- Runs a FastAPI server exposing `POST /haiku` (accepts a `{"topic": "..."}` body)
- Builds a minimal LangGraph state graph with a single `write_haiku` node that calls Azure OpenAI
- Relies on OneAgent to capture the agent's LLM calls as `gen_ai.*` spans

> - Ships an OpenPipeline config (`openpipeline-langgraph.yaml`) that redacts captured messages mentioning secrets, server-side on ingest

Reviewer: could we redact only the secret? 🤔 for example, if it matches a regex
```suggestion
- Ships an OpenPipeline config (`openpipeline-langgraph.yaml`) that redacts messages containing secrets on ingest
```

**Processor matcher** — the matcher field supports a restricted DQL subset: `isNotNull()`, `isNull()`, equality operators (`==`, `!=`), and `AND`/`OR`/`NOT`. Functions like `matchesPhrase()`, `contains()`, and `matches()` are **not** available here for string content checks.

**DQL script** — the script field has a restricted command set: `filter` is not enabled and `contains()` cannot be used as a standalone filter. The correct pattern for conditional redaction is a `fieldsAdd` with an `if()` expression.

`gen_ai.input.messages` is stored as an **array type**, not a plain string, so `contains()` alone does not match against its content. Wrap it with `toString()` first to serialise the array to a string, then apply `contains()`:

```dql
fieldsAdd gen_ai.input.messages = if(contains(toString(gen_ai.input.messages), "secret"), "***REDACTED***", else: gen_ai.input.messages)
```

> Note the required named `else:` parameter — a positional third argument is rejected. Also note that backtick-quoting field names (`` `gen_ai.input.messages` ``) is **not** valid DQL syntax here and will cause a parse error.

Reviewer: ```suggestion
Note the required named `else:` parameter — a positional third argument is rejected. Also note that backtick-quoting field names (`gen_ai.input.messages`) is **not** valid DQL syntax here and will cause a parse error.
```
